[bcr-wdc-key-service] kid-per-ebill to kid-per-red_date

codingpeanut157 opened the pull request:
### **User description**
jumbo-refactor of the key-service that migrates its logic
from a keyset-per-ebill architecture to keyset-per-redemption date.

Previously, the key-service needed to have two separate keysDB, one for
temporary keysets generated during the quoting and minting process and
one for circulating `crsat`s,
We now no longer need such setup as there not longer exists such temporary
state where the keyset is generated for an ebill that is not yet
correctly endorsed to the wildcat.
Each keysetID is linked 1:1 to a redemption date and is enabled and
usable by default.
The minting operations are stored in a separate table and are filled by
a API call by the quote-service.

=== 
[bcr-wdc-quote-service] new keyset logic]

the quote service now exposes a POST method to enable minting for a
given quote ID.
This method is supposed to be called manually from the wildcat dashboard
once the admin has double checked the related ebill endorsement happened correctly

Wildcat minting fees are generated only when the minting is enabled.


## Note
e2e tests to be fixed/extended to the new logic


___

### **PR Type**
Enhancement


___

### **Description**
• **Major architecture refactor**: Migrated key-service from keyset-per-ebill to keyset-per-redemption-date architecture
• **Database schema changes**: Replaced `MintCondition` with `MintOperation` and separated mint operations into dedicated table
• **Service consolidation**: Unified `QuoteKeysRepository` into single `KeysRepository` with Arc-wrapped thread safety
• **API updates**: Added authentication support with `AuthorizationPlugin` for admin endpoints and `/v1/admin/` prefix
• **Keyset generation**: Changed from UUID-based to timestamp-based derivation paths using redemption date components
• **Quote service enhancements**: Added `enable_minting` endpoint and `minting_pubkey` field tracking
• **Client API simplification**: Removed temporary keyset operations and updated to date-based keyset management
• **Configuration improvements**: Updated Docker compose files with proper service dependencies and port mappings
• **Test updates**: Modified test utilities and cases to support new architecture and API changes


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Old: Keyset per E-bill"] --> B["New: Keyset per Redemption Date"]
  C["Separate QuoteKeysRepository"] --> D["Unified KeysRepository with Arc"]
  E["MintCondition"] --> F["MintOperation in separate table"]
  G["UUID-based derivation"] --> H["Timestamp-based derivation"]
  I["Quote-based minting"] --> J["Date-based minting with enable endpoint"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>21 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>surreal.rs</strong><dd><code>Database schema refactor for keyset-per-redemption-date architecture</code></dd></summary>
<hr>

crates/bcr-wdc-key-service/src/persistence/surreal.rs

• Refactored database schema from keyset-per-ebill to <br>keyset-per-redemption-date architecture<br> • Replaced <code>MintCondition</code> with <br><code>MintOperation</code> and separated mint operations into dedicated table<br> • <br>Added new database entry types <code>KeysInfoDBEntry</code> and <code>MintOpDBEntry</code> with <br>conversion methods<br> • Implemented mint operation management methods <br>(<code>store_mintop</code>, <code>load_mintop</code>, <code>list_mintops</code>, <code>update_mintop</code>)


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-ba8ade7df8f929ce375f54a312f93f59224a2beafd836727aa2108c57d32ca26">+411/-363</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>service.rs</strong><dd><code>Service layer refactor for unified keyset management</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-service/src/service.rs

• Replaced separate <code>QuoteKeysRepository</code> with unified <code>KeysRepository</code> <br>using Arc for thread safety<br> • Added <code>get_keyset_id_for_date</code> method to <br>retrieve keysets by redemption date<br> • Refactored minting logic to use <br><code>MintOperation</code> instead of <code>MintCondition</code><br> • Updated mint method to handle <br><code>MintRequest</code> with signature verification and amount tracking


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-0bf63ca2bfa682f3961784061646602648f6450798339852bae9daf95e784e52">+166/-213</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>service.rs</strong><dd><code>Quote service adaptation for new keyset architecture</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/service.rs

• Updated service to use Arc-wrapped trait objects for thread safety<br> • <br>Added <code>enable_minting</code> method to handle minting enablement separately <br>from quote offering<br> • Modified <code>offer</code> method to use <br>keyset-per-redemption-date logic<br> • Updated <code>KeysHandler</code> trait methods <br>to align with new architecture


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-186f7e8fe93bbc7070dc420fa5c734f262aa91e2cc4a5585eecc838222112f36">+101/-86</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Client API updates for date-based keyset management</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-client/src/lib.rs

• Removed quote-based keyset generation methods (<code>pre_sign</code>, <br><code>generate_keyset</code>, <code>enable_keyset</code>)<br> • Added <code>keys_for_expiration</code> and <br><code>new_mint_operation</code> methods for date-based keyset management<br> • Updated <br>mint method to use <code>MintRequest</code> with UUID quote identifier<br> • Simplified <br>client interface by removing temporary keyset operations


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-a8c55995b2dad28ce9375fabcc6c909015b3419a52b5fa6a1c7a4205f25dbf89">+74/-100</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>inmemory.rs</strong><dd><code>In-memory repository consolidation and mint operations support</code></dd></summary>
<hr>

crates/bcr-wdc-key-service/src/persistence/inmemory.rs

• Removed <code>InMemoryQuoteKeyMap</code> and consolidated into unified <br><code>InMemoryKeyMap</code><br> • Added mint operations tracking with separate hash <br>maps for operations and kid references<br> • Implemented mint operation <br>CRUD methods (<code>store_mintop</code>, <code>load_mintop</code>, <code>list_mintops</code>, <code>update_mintop</code>)<br> <br>• Added <code>infos_for_expiration_date</code> method for date-based keyset <br>filtering


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-3f12dfeeadb01ce475140bd0f334bac22948a837c7d6f5ce31a4120e1a8ef560">+117/-156</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Authentication and authorization support for admin endpoints</code></dd></summary>
<hr>

crates/bcr-wdc-ebill-client/src/lib.rs

• Added authentication support with <code>AuthorizationPlugin</code> for authorized <br>endpoints<br> • Updated all admin endpoints to use <code>/v1/admin/</code> prefix and <br>authorization<br> • Added <code>authenticate</code> method for OAuth-style <br>authentication<br> • Modified all API calls to use authorized requests <br>where required


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-a16ec5eb3940ebaff1e6279b5b37f628510942532a0e1720c6b5617e0ad787dd">+70/-28</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>web.rs</strong><dd><code>Web API simplification and mint request handling updates</code>&nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-service/src/web.rs

• Simplified web handlers by removing generic type parameters<br> • <br>Updated mint endpoint to use <code>MintRequest<uuid::Uuid></code> instead of string-based <br>quotes<br> • Removed quote-based signature verification logic from mint <br>handler<br> • Updated all handlers to work with unified <code>Service</code> type


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-02c5aa52715b6e2c505a69f778d2d2ae101cfa3e43d41b471c321b11a362be35">+37/-75</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>admin.rs</strong><dd><code>Treasury service signature storage simplification</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-treasury-service/src/admin.rs

• Removed <code>expiration</code> parameter from <code>store_signatures</code> method call<br> • <br>Updated signature storage to not require expiration timestamp


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-1a696476126e3290a8fea9f62843118cdb79f77c324ead24a867267d7dacaff3">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Refactor key service from quote-based to date-based architecture</code></dd></summary>
<hr>

crates/bcr-wdc-key-service/src/lib.rs

• Removed <code>QuoteKeysRepository</code> type and related generic parameters from <br>service and routes<br> • Changed service to use <code>Arc</code> wrappers for <br>repositories and simplified generic constraints<br> • Updated admin routes <br>to replace quote-based operations with date-based keyset retrieval and <br>mint operations<br> • Removed cashu nut imports in favor of direct cashu <br>imports


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-3f8809d40f9e828f899b297d0eed35852796e26fa96f3c629ac0f4a6dc079589">+31/-63</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>admin.rs</strong><dd><code>Update admin endpoints for date-based keyset management</code>&nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-service/src/admin.rs

• Replaced <code>generate</code> and <code>pre_sign</code> endpoints with <code>get_keyset_for_date</code> <br>and <code>new_mintop</code><br> • Removed <code>enable</code> endpoint and simplified function <br>signatures by removing generic constraints<br> • Updated all admin <br>functions to work with the new <code>Service</code> type without generic parameters


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-f366824ed4999514276e404ad0b38b03fbb4c1768345ebce12638f7be6bfad1f">+57/-91</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>quotes.rs</strong><dd><code>Enhance quote status with minting pubkey tracking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/quotes.rs

• Added <code>minting_pubkey</code> field to quote status variants (<code>Pending</code>, <br><code>Offered</code>, <code>Accepted</code>)<br> • Converted <code>TryFrom</code> implementation to standalone <br><code>convert_to_billinfo</code> function<br> • Enhanced error handling with <br><code>InvalidQuoteStatus</code> replacing <code>QuoteAlreadyResolved</code><br> • Added <code>Display</code> <br>derive to <code>StatusDiscriminants</code> enum


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-03545ff0e2dffaef5dbb0e2990ee74cf84c03a3c10f8f0f069b28ac5734f1ef8">+60/-42</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>factory.rs</strong><dd><code>Migrate keyset generation from UUID to timestamp-based paths</code></dd></summary>
<hr>

crates/bcr-wdc-key-service/src/factory.rs

• Changed keyset generation from UUID-based to timestamp-based <br>derivation paths<br> • Replaced custom <code>generate_mintkeyset</code> function with <br>direct <code>cashu::MintKeySet::generate</code> usage<br> • Updated derivation path <br>calculation to use redemption date timestamp components


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-0936b33fd56c8a54b5936db31eea011621bb1d417e92c43b5751a1bac4ed393a">+42/-68</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>admin.rs</strong><dd><code>Simplify admin service interface and add minting enablement</code></dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/admin.rs

• Simplified function signatures by removing generic type parameters<br> • <br>Added new <code>enable_minting</code> endpoint for quote minting enablement<br> • <br>Updated existing admin functions to work with simplified <code>Service</code> type


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-0ac7a81b04baaf01768d8526f8dceb0754d20adf93d53808f189a4b810752acb">+44/-40</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>web.rs</strong><dd><code>Update web endpoints for minting pubkey integration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/web.rs

• Updated <code>enquire_quote</code> to use <code>minting_pubkey</code> instead of <code>public_key</code><br> • <br>Modified status reply conversion to include <code>minting_pubkey</code> in relevant <br>variants<br> • Simplified function signatures by removing generic type <br>constraints


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-1bd0e6958ea77d8d30de256aa68ee557a88bc615c71586889a98edca3ea18e30">+23/-43</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Refactor service architecture with Arc-wrapped dependencies</code></dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/lib.rs

• Wrapped service dependencies in <code>Arc</code> for shared ownership<br> • <br>Simplified route function by removing generic type parameters<br> • <br>Updated OpenAPI documentation paths and added new <code>enable_minting</code> <br>endpoint


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-165657972ffc04299a11d483cafb1e6a8590ee69e9f7a7170b74a4d93b04b628">+25/-29</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>keys.rs</strong><dd><code>Refactor keys handler interface for date-based operations</code></dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/keys.rs

• Replaced <code>generate</code> method with <code>get_keyset_with_redemption_date</code> and <br><code>add_new_mint_operation</code><br> • Updated <code>sign</code> method to remove quote ID <br>parameter<br> • Modified test utilities to support new keyset retrieval by <br>redemption date


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-d24b03325dacb5f2241120b3f10ef53990566604d941fe8a1dfd7a4ee08098ed">+51/-26</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Expand quote client with comprehensive quote lifecycle methods</code></dd></summary>
<hr>

crates/bcr-wdc-quote-client/src/lib.rs

• Added new methods for quote management: <code>deny</code>, <code>offer</code>, <code>accept_offer</code>, <br><code>reject_offer</code>, <code>cancel_enquiry</code><br> • Updated <code>enquire</code> method parameter from <br><code>mint_pubkey</code> to <code>minting_pubkey</code><br> • Added <code>enable_minting</code> method for <br>authorized quote minting enablement


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-fa0779382859a62530dd708ed21f321792e966dee691bb62cb62946804e28a43">+70/-7</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>wallet.rs</strong><dd><code>Simplify wallet signature storage interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/wallet.rs

• Removed <code>expiration</code> parameter from <code>store_signatures</code> method in both <br>trait and implementation<br> • Updated test utilities to match simplified <br>signature storage interface


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-82a31ebeb1d1b6f5737bc38f6056c3ca659b8a68cd6e6374a8768e9dec0f776a">+7/-11</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>keys.rs</strong><dd><code>Refactor keys API schema for mint operation workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-webapi/src/keys.rs

• Renamed request/response types from quote-based to operation-based <br>naming<br> • Updated field names and types to reflect new mint operation <br>workflow<br> • Replaced enable/generate operations with new mint operation <br>creation


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-188353c6638ec833337a71f3ccf5522a22bb02a6fbfa233224a455a5c7bc1e85">+19/-19</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>quotes.rs</strong><dd><code>Update quotes API schema with minting pubkey support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-webapi/src/quotes.rs

• Renamed <code>public_key</code> field to <code>minting_pubkey</code> in <code>EnquireRequest</code><br> • Added <br><code>minting_pubkey</code> field to status reply variants (<code>Offered</code>, <code>Accepted</code>)<br> • <br>Added new <code>EnableMintingRequest</code> and <code>EnableMintingResponse</code> schemas


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-42911c9e24e041418ffc2f1e92de4be1171ef724642b15b4228d37b60499f496">+11/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Simplify treasury client signature storage interface</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-treasury-client/src/lib.rs

• Removed <code>expiration</code> parameter from <code>store_signatures</code> method call<br> • <br>Simplified signature storage request structure


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-a1ed6e2862956edbaae0bcc2d2035feb1556377205b83a6d249a1cc09e65e3ac">+1/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>15 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>surreal.rs</strong><dd><code>Update database tests for minting pubkey schema changes</code>&nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/persistence/surreal.rs

• Updated test cases to include <code>minting_pubkey</code> field in quote status <br>variants<br> • Modified quote status creation in tests to match new schema <br>requirements


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-e0caddd16b02b3646a69fca2fb0c9f351fcb57dcd7d613a1c750a4e1bd051a0b">+11/-6</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_utils.rs</strong><dd><code>Enhance test utilities with flexible bill generation parameters</code></dd></summary>
<hr>

crates/bcr-wdc-webapi/src/test_utils.rs

• Updated <code>generate_random_bill_enquire_request</code> to accept receiver <br>public key and optional amount<br> • Renamed <code>random_date</code> to <br><code>random_datetime</code> for clarity<br> • Modified function to use receiver public <br>key instead of owner keypair for bill creation


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-e1f74ab6df7fe46ff507cbce007fc16efe9c4c63181b4f6025bea28d976819f3">+13/-15</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>list.rs</strong><dd><code>Update list tests for new bill enquire request format</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-client/tests/list.rs

• Updated test calls to use new <code>generate_random_bill_enquire_request</code> <br>signature with receiver public key<br> • Modified enquire calls to use <br>consistent minting public key parameter


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-03e6c0af2a4326a20b8eabba92723a19a6b632f06a408607af0c7a1d1d8c3058">+17/-13</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>swap.rs</strong><dd><code>Update swap tests for simplified keyset storage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-swap-client/tests/swap.rs

• Updated keyset storage calls to use new async <code>store</code> method without <br>mint condition<br> • Removed <code>MintCondition</code> usage in favor of simplified <br>keyset storage


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-602b5e626d0f9375fb5d8af6d0d36c825b097ff327f03923cd38602570088729">+5/-14</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>checkstate.rs</strong><dd><code>Update checkstate tests for simplified keyset storage</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-swap-client/tests/checkstate.rs

• Updated keyset storage to use new async <code>store</code> method without mint <br>condition parameter<br> • Removed <code>MintCondition</code> import and usage


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-f01234c1bfe54c01a69e6709e88f3d7d3d6d92210a272c64cf293e1af50a32d8">+3/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>ebill.rs</strong><dd><code>Update ebill test utilities for renamed datetime function</code></dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/ebill.rs

• Updated test utility function call from <code>random_date</code> to <br><code>random_datetime</code>


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-8f49a9e1d88ad9c59e7ebeb22dd015f30ef99d14528d174bdbe643e56161c81d">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>lookup.rs</strong><dd><code>Update lookup tests for new bill enquire request format</code>&nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-client/tests/lookup.rs

• Updated test to use new <code>generate_random_bill_enquire_request</code> <br>signature with receiver public key<br> • Modified enquire call to use <br>signing key's public key for minting


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-d1f56eaa033a97db8a88e1fcc6320a8fce933dcbb21105719396a0f567f32082">+10/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>enquire.rs</strong><dd><code>Update enquire tests for new bill enquire request format</code>&nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-client/tests/enquire.rs

• Updated test to use new <code>generate_random_bill_enquire_request</code> <br>signature with receiver public key<br> • Modified enquire call to use <br>signing key's public key for minting


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-c1aa452a491e88b5931fae2864857a2a86487522824636ca1e8ca0c8cbe1ffd6">+10/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>restore.rs</strong><dd><code>Update restore tests for async test server creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-client/tests/restore.rs

• Updated test server creation to use async <code>build_test_server</code> method


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-e4573b3394ae596e7acab3ae1fa1b8d9ba05da6f448d1a80923608cfdf2f9844">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>bills.rs</strong><dd><code>Update ebill client tests for new bill enquire request format</code></dd></summary>
<hr>

crates/bcr-wdc-ebill-client/tests/bills.rs

• Updated test to use new <code>generate_random_bill_enquire_request</code> <br>signature with public key parameter


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-f0b2fa0e421864cafc40b00b275d3c6eb89f6f2b24063000b3f1eca1721c655c">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>keyset_info.rs</strong><dd><code>Update keyset info tests for async test server creation</code>&nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-client/tests/keyset_info.rs

• Updated test server creation to use async <code>build_test_server</code> method


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-6d7a4ef044c176e68d64d29e876ba490b0bd397b9d0e7b7a249fe33983e4f9ab">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>keys.rs</strong><dd><code>Update keys tests for async test server creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-client/tests/keys.rs

• Updated test server creation to use async <code>build_test_server</code> method


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-bbadfe5d8bdd6128700fce128139b13346bba8a8aee4b8ca91d89c5a54a348d0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>list_keyset_info.rs</strong><dd><code>Update list keyset info tests for async test server creation</code></dd></summary>
<hr>

crates/bcr-wdc-key-client/tests/list_keyset_info.rs

• Updated test server creation to use async <code>build_test_server</code> method


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-ebbbd10bd44684454038b931da940cc72dd0d8ae0e57a1593e727a0631fd8f2e">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>cancel.rs</strong><dd><code>Update cancel tests for renamed method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-quote-client/tests/cancel.rs

• Updated test to use renamed <code>cancel_enquiry</code> method instead of <code>cancel</code>


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-8d894fc4076b3af42631c49af8068695d9d222436fffbc73a3ac5058a8d7b3de">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>list_keys.rs</strong><dd><code>Update list keys tests for async test server creation</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-client/tests/list_keys.rs

• Updated test server creation to use async <code>build_test_server</code> method


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-ec05b05dc6bf7fa8a1e0b67563b8a71c20a7da8be2f8662c043a23531407e7ac">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>keys.rs</strong><dd><code>Remove UUID-based key derivation utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-utils/src/keys.rs

• Removed <code>extend_path_from_uuid</code> function as UUID-based path generation <br>is no longer needed<br> • Cleaned up imports related to UUID and BIP32 <br>path manipulation


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-88918b1bb74f41ff3324615d303d1c32ed912a92eeefe25d47604cf2b3a362c8">+1/-33</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>devtool.rs</strong><dd><code>Update devtool for new bill enquire request format</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-webapi/src/bin/devtool.rs

• Updated devtool to use new <code>generate_random_bill_enquire_request</code> <br>signature with public key parameter


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-df21ec837a02726f23714c89a58bfc9afeb0f0e568a977a8907f7bed7f32c477">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Add admin prefix to ebill service administrative routes</code>&nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-ebill-service/src/lib.rs

• Updated all route paths to include <code>/admin/</code> prefix for administrative <br>endpoints<br> • Reorganized routes to clearly separate admin functionality <br>from public endpoints


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-bca71cbba9ed5622ab5e4c3fa1910a48a033a88332b9531db05bd4869694b8ce">+18/-12</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>mod.rs</strong><dd><code>Conditionally compile inmemory persistence for tests only</code></dd></summary>
<hr>

crates/bcr-wdc-key-service/src/persistence/mod.rs

• Added conditional compilation for inmemory module to only include in <br>test builds


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-34b7cf60e8e8ea099119b79812adae9f6944c4c0a4bb8996feb9fd78da184710">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>common-services.yml</strong><dd><code>Improve service dependencies and remove unnecessary port exposures</code></dd></summary>
<hr>

common-services.yml

• Removed exposed ports from internal services (ebpp, keycloak, <br>surrealdb)<br> • Added ebill-service dependency to quote-service<br> • Updated <br>bff-wallet-service dependencies to include health checks


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-feef9361dee340026246ae79275bc3b63be23a97aef0579ef9042376439c4137">+6/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>docker-compose-e2e.yml</strong><dd><code>Optimize e2e test environment service dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docker-compose-e2e.yml

• Added port exposure for surrealdb and keycloak in e2e environment<br> • <br>Simplified e2e-tests service dependencies to only essential services


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-a7288a5d1665494bd3e308cf33c299712053eb89448da79918583891af8b2a23">+5/-11</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>docker-compose-local.yml</strong><dd><code>Fix service reference and add port mappings for local development</code></dd></summary>
<hr>

docker-compose-local.yml

• Fixed typo in swap-service reference and added port exposures for <br>local development<br> • Added port mappings for surrealdb, <br>bff-dashboard-service, and bff-wallet-service


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-aed91aa2667654a05077a027349bdccf0caa1bfefddc15a397fc00f1ee8385d3">+7/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>envoy.yaml</strong><dd><code>Simplify envoy routing configuration for dashboard service</code></dd></summary>
<hr>

docker/bff-dashboard-service/envoy.yaml

• Removed prefix rewrite rules for identity and bill payment status <br>routes<br> • Simplified route configuration to direct cluster routing


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-42eb1a6f496726d980f9f5be9b9b7e5213440f12b3ee73b5a3d98907903a9b7c">+2/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Error handling</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>error.rs</strong><dd><code>Improve quote error handling with status-specific validation</code></dd></summary>
<hr>

crates/bcr-wdc-quote-service/src/error.rs

• Replaced <code>QuoteAlreadyResolved</code> error with more specific <br><code>InvalidQuoteStatus</code> error<br> • Updated error response mapping to handle <br>new status validation errors


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-ba4223300771963ac9bac2c3a6d4859b9e260cff57010253b4ef8f8d51f641ef">+9/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>error.rs</strong><dd><code>Add internal error handling to key service</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-service/src/error.rs

• Added new <code>Internal</code> error variant for internal server errors<br> • <br>Updated error response mapping to handle internal errors with 500 <br>status code


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-1ddc03038fa2a6c659a03a250ed71bf55253dabf322d0caad60ec76a7e73da7c">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>signatures.rs</strong><dd><code>Fix variable naming in blind signature validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-utils/src/signatures.rs

• Fixed variable naming from <code>unique_proofs</code> to <code>unique_blinds</code> for <br>clarity in blind signature validation


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-3a6b632ae0793c851aa4bac969ce483cc94c1bdc2954c1af74d159ae2de91e8f">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>Cargo.toml</strong><dd><code>Add authorization feature support to key client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-key-client/Cargo.toml

• Added <code>authorized</code> feature dependency on <code>bcr-wdc-utils/auth</code><br> • Added <br>dev-dependency with test-utils and authorized features


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-f0d2d228315ef718f33c9cc1fc49a4002ccd011ae5e7aaae8415d0f7d11f0f32">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Cargo.toml</strong><dd><code>Add authorization feature support to ebill client</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

crates/bcr-wdc-ebill-client/Cargo.toml

• Added <code>authorized</code> feature with bcr-wdc-utils dependency<br> • Added <br>dev-dependencies for testing with authorization features


</details>


  </td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-eca02ddaff04e57fc896ad2ec840dbf27c58b1e0e4318e4fc35913dc519fcc9c">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Additional files</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>pre_sign.rs</strong></td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-2b05713a9668b140f4313ba736cbd54d1582040fb1d71de4cd356c0a0b3d0a7f">+0/-28</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Cargo.toml</strong></td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-4e256eed7ead076418f608369ce678e64c1a2ee653ee3518b1f27a22b3c502b0">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Cargo.toml</strong></td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-7953f8a4c41812869b8fde0f8bf44078cf538d9bc28115d3d37e77e77124cdfe">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mod.rs</strong></td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-fb8e907942c08710bc61e846f07cd964a3a620ff03e6df23ec32a299dc7a6558">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>signatures.rs</strong></td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-56eeabfa487aff0407262229cab0d9d44a5bb537e1a40ff0e41410dc1b9ceac9">+0/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>config.toml</strong></td>
  <td><a href="https://github.com/BitcreditProtocol/Wildcat/pull/295/files#diff-ad1172e87dd1b09e7745ee6029056d0ced2d5322feea8e908065d0d1774f726a">+2/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

zupzup commented on crates/bcr-wdc-quote-client/src/lib.rs:
```diff
@@ -169,4 +169,16 @@ impl QuoteClient {
+        let url = self
+            .base
+            .join(&format!("/v1/admin/credit/quote/enable_mint/{qid}"))
+            .expect("list relative path");
```
"enable mint relative path"

zupzup commented on crates/bcr-wdc-quote-service/src/keys.rs:
```diff
@@ -27,27 +27,36 @@ impl KeysRestHandler {
 #[async_trait]
 impl KeysHandler for KeysRestHandler {
-    async fn generate(
+    async fn get_keyset_for_with_expiration_date(
```
for_with? 🤔 Sounds a bit strange.

for_expiration_date, or with_expiration_date?

zupzup commented on crates/bcr-wdc-key-service/src/factory.rs:
```diff
@@ -28,72 +27,51 @@ impl Factory {
+            btc32::ChildNumber::from_hardened_idx(u16_1).unwrap(),
+            btc32::ChildNumber::from_hardened_idx(u16_0).unwrap(),
+        ];
+        dbg!(tstamp, u16_3, u16_2, u16_1, u16_0, extension);
```
leftover?

zupzup commented on crates/bcr-wdc-key-service/src/factory.rs:
```diff
@@ -28,72 +27,51 @@ impl Factory {
-        let set = generate_mintkeyset(
-            self.master,
-            Self::MAX_ORDER,
+    pub fn generate(&self, expire: TStamp) -> keys::KeysetEntry {
```
Is this process documented somewhere? (the steps)

codingpeanut157 replied:
You mean the whole quoting system?
I should probably about the internal documentation :thinking:

zupzup replied:
No, no just concretely the algorithm in the `generate` method, since there was an explaining comment in the `generate_mintkeyset` function that referenced cashu

zupzup commented on crates/bcr-wdc-key-service/src/service.rs:
```diff
@@ -198,88 +120,91 @@ where
+        if signature_verification.is_err() {
             return Err(Error::InvalidMintRequest(String::from(
-                "keyset already minted",
+                "signature verifaction failed",
```
verification*

zupzup commented on crates/bcr-wdc-key-service/src/persistence/surreal.rs:
```diff
@@ -38,296 +41,265 @@ fn convert_from(ke: KeysetEntry, condition: MintCondition) -> KeysDBEntry {
+            .map_err(|e| Error::KeysRepository(anyhow!(e)))?
+            .take(0)
+            .map_err(|e| Error::KeysRepository(anyhow!(e)))?;
+        if let Some(before) = before {
```
I don't get this 😅 

So you update and set minted to the given minted amount, returning the value as it was before the update.
Then you check if the value before the update was smaller than the given value and log that, but it's not an error. Correct?

Is this a validation? Or a precondition that should hold?

In any case, I think it might be good to add a comment why this is important to only log this and not e.g. error out 😅

codingpeanut157 replied:
You are right, I should probably implement some logic that prevents `minted` to "overflow" the `target` value

zupzup reviewed (approved):
LGTM 👍

codingpeanut157 committed:
[bcr-wdc-webapi] new wire types for keyset-per-day

refactor of test-utils functions

codingpeanut157 committed:
[bcr-wdc-key-service] refactor to keyset-per-day architecture

jumbo-refactor of the key-service that migrates its logic
from a keyset-per-ebill architecture to keyset-per-redemption date.

Previously, the key-service needed to have two separate keysDB, one for
temporary keysets generated during the quoting and minting process and
one for circulating `crsat`s,
We now no longer need such setup as there not longer exists such temporary
state where the keyset is generated for an ebill that is not yet
correctly endorsed to the wildcat.
Each keysetID is linked 1:1 to a redemption date and is enabled and
usable by default.
The minting operations are stored in a separate table and are filled by
a API call by the quote-service.

codingpeanut157 committed:
[bcr-wdc-key-client] refactor to adapt to new key-service

codingpeanut157 committed:
[bcr-wdc-utils] minor changes

remove unused functions, better variable names

codingpeanut157 committed:
[bcr-wdc-ebill-service] admin endpoint behind /admin section

codingpeanut157 committed:
[bcr-wdc-quote-client] refactor to adapt to new quote-service

codingpeanut157 committed:
[bcr-wdc-treasury-service] refactor to adapt to new wire types

codingpeanut157 committed:
[bcr-wdc-treasury-client] refactor to adapt to new treasury-service

codingpeanut157 committed:
[bcr-wdc-swap-client] refactor to adapt to new key-client

it concerns only tests

codingpeanut157 committed:
[bcr-wdc-ebill-client] add authorized features

all ebill-client endpoints require authorization.
This commit integrates the bcr-wdc-utils::client authorization plugin
into the EBillClient

codingpeanut157 committed:
[bcr-wdc-quote-service] new keyset logic

the quote service now exposes a POST method to enable minting for a
given quote ID.
This method is supposed to be called manually from the wildcat dashboard
once the admin has double checked the related ebill endorsement happened correctly

Wildcat minting fees are generated only when the minting is enabled.

codingpeanut157 committed:
[docker] better dependency description

ports forwarding in docker-compose rather than common-services

codingpeanut157 commented:
Fixing the e2e is taking too long, due to a mix of carelessness and hiccups due to wildcat repo visibility changes

I will leave it for a separate PR

stefanbitcr commented on docker-compose-local.yml:
```diff
@@ -16,7 +16,7 @@ services:
     extends:
         file: common-services.yml
-        service: swap-service
+        service: swap-service0
```
This one isn't in common-services (assume it was for testing)

codingpeanut157 committed:
[bcr-wdc-e2e-tests] make e2e tests use our clients

still non-working tests

codingpeanut157 committed:
[github] add step to clean-up the runner storage
